File: sdk/typescript/src/resources.ts
/**
 * @license
 * Copyright 2026 cofy-x
 * SPDX-License-Identifier: Apache-2.0
 */

import { SandboxValidationError } from "./errors/index.js";

export type ResourceQuantity = string | number;

export interface ResourceOptions {
  requestCpu?: ResourceQuantity;
  requestMemory?: ResourceQuantity;
  requestEphemeralStorage?: ResourceQuantity;
  limitCpu?: ResourceQuantity;
  limitMemory?: ResourceQuantity;
  limitEphemeralStorage?: ResourceQuantity;
}

export function buildResourceSpec(options: ResourceOptions): Record<string, unknown> | undefined {
  const requests = resourceQuantity(
    "requestCpu",
    options.requestCpu,
    "requestMemory",
    options.requestMemory,
    "requestEphemeralStorage",
    options.requestEphemeralStorage,
  );
  const limits = resourceQuantity(
    "limitCpu",
    options.limitCpu,
    "limitMemory",
    options.limitMemory,
    "limitEphemeralStorage",
    options.limitEphemeralStorage,
  );
  if (requests === undefined && limits === undefined) {
    return undefined;
  }
  return { requests, limits };
}

function resourceQuantity(
  cpuName: string,
  cpuValue: ResourceQuantity | undefined,
  memoryName: string,
  memoryValue: ResourceQuantity | undefined,
  ephemeralStorageName: string,
  ephemeralStorageValue: ResourceQuantity | undefined,
): Record<string, number> | undefined {
  const cpu = parseCpuQuantity(cpuName, cpuValue);
  const memory = parseMemoryQuantity(memoryName, memoryValue);
  const ephemeralStorage = parseMemoryQuantity(ephemeralStorageName, ephemeralStorageValue);
  if (cpu <= 0 && memory <= 0 && ephemeralStorage <= 0) {
    return undefined;
  }
  return {
    cpu_milli: cpu,
    memory_bytes: memory,
    ephemeral_storage_bytes: ephemeralStorage,
  };
}

function parseCpuQuantity(name: string, value: ResourceQuantity | undefined): number {
  const quantity = resourceQuantityText(name, value);
  if (quantity === "") {
    return 0;
  }
  if (quantity.startsWith("-")) {
    throw new SandboxValidationError(`${name} must be non-negative`);
  }
  if (quantity.endsWith("m")) {
    return parseScaledDecimal(name, quantity.slice(0, -1), 1n);
  }
  return parseScaledDecimal(name, quantity, 1000n);
}

function parseMemoryQuantity(name: string, value: ResourceQuantity | undefined): number {
  const quantity = resourceQuantityText(name, value);
  if (quantity === "") {
    return 0;
  }
  if (quantity.startsWith("-")) {
    throw new SandboxValidationError(`${name} must be non-negative`);
  }
  const [number, factor] = splitMemoryQuantity(quantity);
  return parseScaledDecimal(name, number, factor);
}

function resourceQuantityText(name: string, value: ResourceQuantity | undefined): string {
  if (value === undefined) {
    return "";
  }
  if (typeof value === "number") {
    if (!Number.isFinite(value)) {
      throw new SandboxValidationError(`${name} must be a resource quantity`);
    }
    return String(value);
  }
  return value.trim();
}

function splitMemoryQuantity(value: string): [string, bigint] {
  const units: Array<[string, bigint]> = [
    ["tib", 1024n ** 4n],
    ["gib", 1024n ** 3n],
    ["mib", 1024n ** 2n],
    ["kib", 1024n],
    ["ti", 1024n ** 4n],
    ["gi", 1024n ** 3n],
    ["mi", 1024n ** 2n],
    ["ki", 1024n],
    ["tb", 1000n ** 4n],
    ["gb", 1000n ** 3n],
    ["mb", 1000n ** 2n],
    ["kb", 1000n],
    ["b", 1n],
  ];
  const lower = value.toLowerCase();
  for (const [suffix, factor] of units) {
    if (lower.endsWith(suffix)) {
      return [value.slice(0, -suffix.length).trim(), factor];
    }
  }
  return [value, 1n];
}

function parseScaledDecimal(name: string, value: string, scale: bigint): number {
  const quantity = value.trim();
  if (quantity === "") {
    throw new SandboxValidationError(`${name} is required`);
  }
  const match = quantity.match(/^(\+)?(\d+|\d*\.\d+)$/);
  if (match === null) {
    throw new SandboxValidationError(`${name} must be a decimal number`);
  }
  const [wholePart, fractionPart = ""] = quantity.replace(/^\+/, "").split(".");
  const whole = wholePart === "" ? "0" : wholePart;
  const numerator = BigInt(`${whole}${fractionPart}`) * scale;
  const denominator = 10n ** BigInt(fractionPart.length);
  if (numerator % denominator !== 0n) {
    throw new SandboxValidationError(`${name} must resolve to a whole unit`);
  }
  const parsed = numerator / denominator;
  if (parsed > BigInt(Number.MAX_SAFE_INTEGER)) {
    throw new SandboxValidationError(`${name} is too large`);
  }
  return Number(parsed);
}

File: sdk/go/resources.go
package axernsdk

import (
	"math/big"
	"strconv"
	"strings"
)

// ResourceQuantity is a user-facing resource quantity. String literals such as
// "500m", "512Mi", and "512MiB" can be assigned directly; constructors below
// make numeric quantities explicit without weakening the public API to any.
type ResourceQuantity string

// CPUCores formats CPU cores as a resource quantity.
func CPUCores(cores int64) ResourceQuantity {
	return ResourceQuantity(strconv.FormatInt(cores, 10))
}

// CPUMilli formats milli CPU as a resource quantity.
func CPUMilli(milli int64) ResourceQuantity {
	return ResourceQuantity(strconv.FormatInt(milli, 10) + "m")
}

// MemoryBytes formats memory bytes as a resource quantity.
func MemoryBytes(bytes int64) ResourceQuantity {
	return ResourceQuantity(strconv.FormatInt(bytes, 10))
}

// EphemeralStorageBytes formats node-local ephemeral storage bytes as a resource quantity.
func EphemeralStorageBytes(bytes int64) ResourceQuantity {
	return ResourceQuantity(strconv.FormatInt(bytes, 10))
}

func parseCPUQuantity(field string, quantity ResourceQuantity) (int64, error) {
	value := strings.TrimSpace(string(quantity))
	if value == "" {
		return 0, nil
	}
	if strings.HasPrefix(value, "-") {
		return 0, positiveQuantityError(field)
	}
	if strings.HasSuffix(value, "m") {
		return parseScaledResourceDecimal(field, strings.TrimSuffix(value, "m"), 1)
	}
	return parseScaledResourceDecimal(field, value, 1000)
}

func parseMemoryQuantity(field string, quantity ResourceQuantity) (int64, error) {
	value := strings.TrimSpace(string(quantity))
	if value == "" {
		return 0, nil
	}
	if strings.HasPrefix(value, "-") {
		return 0, positiveQuantityError(field)
	}
	number, factor, ok := splitMemoryQuantity(value)
	if !ok {
		return 0, &ValidationError{Field: field, Message: "unsupported unit"}
	}
	return parseScaledResourceDecimal(field, number, factor)
}

func splitMemoryQuantity(value string) (string, int64, bool) {
	units := []struct {
		suffix string
		factor int64
	}{
		{suffix: "tib", factor: 1024 * 1024 * 1024 * 1024},
		{suffix: "gib", factor: 1024 * 1024 * 1024},
		{suffix: "mib", factor: 1024 * 1024},
		{suffix: "kib", factor: 1024},
		{suffix: "ti", factor: 1024 * 1024 * 1024 * 1024},
		{suffix: "gi", factor: 1024 * 1024 * 1024},
		{suffix: "mi", factor: 1024 * 1024},
		{suffix: "ki", factor: 1024},
		{suffix: "tb", factor: 1000 * 1000 * 1000 * 1000},
		{suffix: "gb", factor: 1000 * 1000 * 1000},
		{suffix: "mb", factor: 1000 * 1000},
		{suffix: "kb", factor: 1000},
		{suffix: "b", factor: 1},
	}
	lower := strings.ToLower(value)
	for _, unit := range units {
		if strings.HasSuffix(lower, unit.suffix) {
			return strings.TrimSpace(value[:len(value)-len(unit.suffix)]), unit.factor, true
		}
	}
	return value, 1, true
}

func parseScaledResourceDecimal(field, value string, scale int64) (int64, error) {
	value = strings.TrimSpace(value)
	if value == "" {
		return 0, &ValidationError{Field: field, Message: "is required"}
	}
	value = strings.TrimPrefix(value, "+")
	parts := strings.Split(value, ".")
	if len(parts) > 2 || parts[0] == "" && len(parts) == 1 {
		return 0, &ValidationError{Field: field, Message: "must be a decimal number"}
	}
	whole := parts[0]
	fraction := ""
	if len(parts) == 2 {
		fraction = parts[1]
	}
	if whole == "" {
		whole = "0"
	}
	if fraction == "" && strings.HasSuffix(value, ".") {
		return 0, &ValidationError{Field: field, Message: "must be a decimal number"}
	}
	digits := whole + fraction
	for _, r := range digits {
		if r < '0' || r > '9' {
			return 0, &ValidationError{Field: field, Message: "must be a decimal number"}
		}
	}
	numerator, ok := new(big.Int).SetString(digits, 10)
	if !ok {
		return 0, &ValidationError{Field: field, Message: "must be a decimal number"}
	}
	numerator.Mul(numerator, big.NewInt(scale))
	denominator := new(big.Int).Exp(big.NewInt(10), big.NewInt(int64(len(fraction))), nil)
	quotient, remainder := new(big.Int), new(big.Int)
	quotient.QuoRem(numerator, denominator, remainder)
	if remainder.Sign() != 0 {
		return 0, &ValidationError{Field: field, Message: "must resolve to a whole unit"}
	}
	if !quotient.IsInt64() {
		return 0, &ValidationError{Field: field, Message: "is too large"}
	}
	return quotient.Int64(), nil
}

func positiveQuantityError(field string) error {
	return &ValidationError{Field: field, Message: "must be non-negative"}
}
